### PR TITLE
More details on collection expression conversions

### LINKED ISSUE #39157: Collection expression [] and Array.Empty
### Type of issue

Missing information

### Description

I see collection expressions suggested as a replacement for `Array.Empty` in some examples around, but no explicit mention that this new expression will use `Array.Empty` for empty collections (`[]`). Is this memory optimisation considered when resolving such an expression to compiled code? A confirmation in the documentation would be appreciated.

BTW, this page also shows how `[]` can be converted to `Span`. But Span is just a view on some other collection, right? So what is the underlying storage type in this case? The whole documentation here looks more like a demonstration of magic than a technical explanation to me. I could decompile and look it up, but then I don't need the documentation anymore...

### Page URL

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/collection-expressions

### Content source URL

https://github.com/dotnet/docs/blob/main/docs/csharp/language-reference/operators/collection-expressions.md

### Document Version Independent Id

49349466-94fd-00b5-93da-e4200d9f9ec8

### Article author

@BillWagner

### Metadata

* ID: 63733f9f-0896-c948-1a7e-58fb428c11c0 
* Service: **dotnet-csharp**
* Sub-service: **lang-reference**

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #39157

Add examples on how the compiler converts a collection expression to an underlying collection.

The specification gives the compiler latitude about how to implement any conversion from a collection expression to a sequence. For that reason, I used a couple examples rather than write rules based on the current implementation.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/csharp/language-reference/operators/collection-expressions.md](https://github.com/dotnet/docs/blob/0f4e20108df5f76e2e1e86150ff815de190d7424/docs/csharp/language-reference/operators/collection-expressions.md) | [Collection expressions - C# language reference](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/collection-expressions?branch=pr-en-us-39581) |

### CODE REVIEW

**Comment by BillWagner** [comment]
ping @adegeo for review.

File: docs/csharp/language-reference/operators/collection-expressions.md [hunk 1]
```diff
@@ -39,6 +39,8 @@ A *collection expression* can be converted to different collection types, includ
...
 
+The [Feature specification](~/_csharplang/proposals/csharp-12.0/collection-expressions.md) specifies how the compiler determines how to convert a collection expression to a target collection. The specification provides flexibility for the compiler to generate the most performant conversion to the target expression. For example, the empty collection expression, `[]`, can be realized as <xref:System.Array.Empty%60%601?displayProperty=nameWithType> if the target won't be modified after initialization. When the target is a <xref:System.Span%601?displayProperty=nameWithType> or <xref:System.ReadOnlySpan%601?displayProperty=nameWithType>, the storage may be stack allocated.
```
Comment (adegeo) [inline docs/csharp/language-reference/operators/collection-expressions.md:42]: Total nit. Two things:

1. Should **Feature** be capitalized?
2. As a reader about this subject, I don't necessarially care too much as to what the feature spec says things should operate. I'm more concerned about how they operate. I would suggest reversing the paragraph. Lead with how it works and follow up with more information bits related to the feature spec.
3. As a reader coming in, what's the "feature specification" anyhow? The C# language specification?

**Review by adegeo — APPROVED** [review]
_Disabled auto merge_

@BillWagner  One thing to consider and throw away.

File: docs/csharp/language-reference/operators/collection-expressions.md [hunk 2]
```diff
@@ -39,6 +39,8 @@ A *collection expression* can be converted to different collection types, includ
...
 
+The compiler uses static analysis to determine the most performant way to create the collection declared with a collection expression. For example, the empty collection expression, `[]`, can be realized as <xref:System.Array.Empty%60%601?displayProperty=nameWithType> if the target won't be modified after initialization. When the target is a <xref:System.Span%601?displayProperty=nameWithType> or <xref:System.ReadOnlySpan%601?displayProperty=nameWithType>, the storage may be stack allocated. The [Feature specification](~/_csharplang/proposals/csharp-12.0/collection-expressions.md) specifies the rules the compiler must follow.
```
Comment (BillWagner) [inline docs/csharp/language-reference/operators/collection-expressions.md:42]:
```suggestion
The compiler uses static analysis to determine the most performant way to create the collection declared with a collection expression. For example, the empty collection expression, `[]`, can be realized as <xref:System.Array.Empty%60%601?displayProperty=nameWithType> if the target won't be modified after initialization. When the target is a <xref:System.Span%601?displayProperty=nameWithType> or <xref:System.ReadOnlySpan%601?displayProperty=nameWithType>, the storage may be stack allocated. The [collection expressions feature specification](~/_csharplang/proposals/csharp-12.0/collection-expressions.md) specifies the rules the compiler must follow.
```

**Comment by adegeo** [comment]
@BillWagner Much better in my opinion! Nice job!

**Pull request merged**